platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/ExecutionEngine.java

```diff
@@ -65,6 +66,9 @@ interface Result {
 
         CachingState getCachingState();
 
+        // TODO Parametrize UnitOfWork with this type
+        <T> Try<T> resolveOutputFromWorkspaceAs(Class<T> type);
```

Comment:
As for `T` instead of `Try`: we need the `Try` for the identity cache, as we are keeping track of failed executions in the identity cache.

I like the simpler name as here the caller doesn't need to know about there being different workspaces.

And yes, we should parameterize `Result` as well as `UnitOfWork` with the type of result to be produced. There's a TODO about doing this right above. However, that's a bigger piece of work that I wasn't ready to take on as part of this PR.